lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,744 @@
+    llvm::DenseSet<Expr *> &cycleBreaker) {
+  // Insert the expr into the set of seen ones. If it's already there, break the
+  // cycle.
+  if (!cycleBreaker.insert(expr).second) {
+    LLVM_DEBUG(llvm::dbgs() << "Detected cycle in " << expr << "\n");
+    return;
+  }
+
+  // If we have already visited this node, we're all good. Otherwise lookup
+  // potential subexprs.
+  if (!dfs.count(expr)) {
+    TypeSwitch<Expr *>(expr)
+        .template Case<VarExpr>([&](auto *var) {
+          auto it = constraints.find(var);
+          if (it != constraints.end())
+            gatherDepthFirstExprs(it->second, dfs, cycleBreaker);
+        })
+        .template Case<AddExpr, MaxExpr, MinExpr>([&](auto *expr) {
+          gatherDepthFirstExprs(expr->lhs, dfs, cycleBreaker);
+          gatherDepthFirstExprs(expr->rhs, dfs, cycleBreaker);
+        });
+    dfs.insert(expr);
+  }
+
+  // Remove the cycle tombstone.
+  cycleBreaker.erase(expr);
+}
+
+//===----------------------------------------------------------------------===//
+// Inference Constraint Problem Mapping
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// A helper class which maps the types and operations in a design to a set of
+/// variables and constraints to be solved later.
+class InferenceMapping {
+public:
+  InferenceMapping(ConstraintSolver &solver) : solver(solver) {}
+
+  void map(CircuitOp op);
+  void map(FModuleOp op);
+  void mapOperation(Operation *op);
+
+  Expr *declareVars(Value value);
+  Expr *declareVars(Type type);
+  Expr *declareVars(FIRRTLType type);
+
+  void unifyTypes(Expr *larger, Expr *smaller);
+
+  Expr *getExpr(Value value);
+  Expr *getExprOrNull(Value value);
+  void setExpr(Value value, Expr *expr);
+
+private:
+  /// The constraint solver into which we emit variables and constraints.
+  ConstraintSolver &solver;
+
+  /// The constraint exprs for each result type of an operation.
+  // TODO: This should actually not map to `Expr *` directly, but rather a
+  // view class that can represent aggregate exprs for bundles/arrays as well.
+  DenseMap<Value, Expr *> opExprs;
+};
+
+} // namespace
+
+void InferenceMapping::map(CircuitOp op) {
+  // llvm::errs() << "Mapping circuit " << op << "\n";
+  // op.walk([this](Operation *op) { map(op); });
+  for (auto &op : *op.getBody()) {
+    if (auto module = dyn_cast<FModuleOp>(&op))
+      map(module);
+  }
+}
+
+void InferenceMapping::map(FModuleOp module) {
+  // Ensure we have constraint variables for the module ports.
+  for (auto arg : module.getArguments())
+    declareVars(arg);
+
+  // Go through operations, creating type variables for results, and generating
+  // constraints.
+  for (auto &op : *module.getBodyBlock()) {
+    mapOperation(&op);
+  }
+}
+
+void InferenceMapping::mapOperation(Operation *op) {
+  TypeSwitch<Operation *>(op)
+      .Case<ConstantOp, WireOp>([&](auto op) { declareVars(op.getResult()); })
```

Comment:
As @drom pointed out in another discussion, `ConstantOp` should not create a free variable in the solver, but just produce the `clog2` of the constant literal.